app/services/foreman/plugin.rb

```diff
@@ -396,5 +417,47 @@ def extend_template_helpers_by_module(mod)
       end
       allowed_template_helpers(*(mod.public_instance_methods - Module.public_instance_methods))
     end
+
+    def add_manager_roles(resources, opts)
+      manager = Role.find_by :name => "Manager"
+      return if !manager || resources.empty?
+      add_permissions manager, resources, opts
+    end
+
+    def add_viewer_roles(resources, opts)
+      viewer = Role.find_by :name => "Viewer"
+      opts[:condition] = "name LIKE :name"
+      opts[:condition_hash] = { :name => "view_%" }
+      return if !viewer || resources.empty?
+      add_permissions viewer, resources, opts
+    end
+
+    def add_permissions(role, resources, opts)
+      to_add = resources.map { |resource| new_permissions_for role, resource, opts }.reduce(:+)
+      role.add_permissions! to_add.map(&:name) if to_add.any?
+    end
+
+    def new_permissions_for(role, resource, opts)
+      condition_hash = { :resource_type => resource }
+      condition_hash.merge! opts[:condition_hash] if opts[:condition_hash]
+      conditions = "resource_type = :resource_type"
+      conditions << (" AND " + opts[:condition]) if opts[:condition]
+      all_permissions = Permission.where(conditions, condition_hash)
+      all_permissions.reject do |permission|
+        permission_already_included? role, permission, opts
+      end
+    end
+
+    def permission_already_included?(role, permission, opts)
+      opts[:except] ||= []
+      opts[:except].include?(permission.name.to_sym) || role.permissions.where(:name => permission.name).any?
+    end
+
+    def add_permissions_to_role(role, permission_names)
+      to_add = permission_names.reject do |name|
+        role.permissions.find_by :name => name
+      end
+      role.add_permissions! to_add if to_add.any?
+    end
```

Comment:
I will simplify my concerns to just single item. Currently there is `default_roles` that provides a hash of roles defined by plugin and list of permissions. Can you deprecate this hash and incorporate it into the new API so every plugin exposes all its roles and permissions via some method? Then modify the `plugin:validate_roles` rake task, it is still useful when users play with roles and does not restart Rails too often. The default permissions exposed are also useful for easier permission testing, so this is useful API feature that I am missing here.